crates/protocol/src/var_int.rs

```diff
@@ -0,0 +1,109 @@
+use std::{
+    io::{Cursor, Read, Write},
+    num::TryFromIntError,
+};
+
+use byteorder::ReadBytesExt;
+use thiserror::Error;
+
+use crate::io::{Decoder, Encoder};
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub struct VarInt(pub i32);
+
+impl VarInt {
+    pub const MAX_SIZE: usize = 5;
+
+    pub const fn written_size(self) -> usize {
+        match self.0 {
+            0 => 1,
+            n => (31 - n.leading_zeros() as usize) / 7 + 1,
+        }
+    }
+
+    pub fn decode_partial(mut r: impl Read) -> Result<i32, VarIntDecodeError> {
```

Comment:
This method should return

```suggestion
    pub fn decode_partial(mut r: impl Read) -> Result<Self, VarIntDecodeError> {
```

which is the more expected behaviour.